File: backend/app/schemas/position.py
from pydantic import BaseModel
from typing import Optional
from datetime import date, datetime


class PositionBase(BaseModel):
    portfolio_code: str
    product_code: str
    market: Optional[str] = None
    platform_code: Optional[str] = None
    shares: Optional[float] = None
    frozen_shares: Optional[float] = None
    cost_price: Optional[float] = None
    unit_price: Optional[float] = None
    market_value: Optional[float] = None
    cash_amount: Optional[float] = None
    frozen_amount: Optional[float] = None
    asset_type: Optional[str] = None
    snapshot_date: date


class PositionCreate(PositionBase):
    pass


class PositionUpdate(BaseModel):
    platform_code: Optional[str] = None
    shares: Optional[float] = None
    frozen_shares: Optional[float] = None
    cost_price: Optional[float] = None
    unit_price: Optional[float] = None
    market_value: Optional[float] = None
    cash_amount: Optional[float] = None
    frozen_amount: Optional[float] = None
    snapshot_date: Optional[date] = None


class PositionResponse(PositionBase):
    id: int
    created_at: Optional[datetime] = None
    # 读侧派生字段（issue：前端持仓表产品名称/盈亏/收益率三列）：
    # product_name 来自 product 表；盈亏 = market_value − shares×cost_price（仅净值型资产）
    product_name: Optional[str] = None
    profit_loss: Optional[float] = None
    profit_loss_percent: Optional[float] = None
    # 读侧派生字段（issue #99）：asset_name 来自 asset_classification（聚合展示短名目）；
    # daily_profit 为当日收益（首个快照日 / IN_TRANSIT 在途行 → None）；
    # is_qdii 来自 product 表（QDII 按 T-1 净值估值，前端提示日收益滞后一天）
    asset_name: Optional[str] = None
    daily_profit: Optional[float] = None
    is_qdii: Optional[bool] = None

    class Config:
        from_attributes = True


class CashPositionUpdate(BaseModel):
    """非净值型资产（现金）更新请求"""
    cash_amount: float
    platform_code: str  # 必填：平台代码
    update_date: Optional[date] = None  # 可选，默认为今天

File: frontend/src/types/position.ts
export interface Position {
  id: number;
  portfolio_code: string;
  product_code: string;
  product_name?: string;
  market?: string;
  platform_code?: string;
  platform_name?: string | null;
  shares?: number;
  frozen_shares?: number;
  cost_price?: number;
  unit_price?: number;
  market_value?: number;
  cash_amount?: number;
  frozen_amount?: number;
  profit_loss?: number;
  profit_loss_percent?: number;
  // 读侧派生（issue #128）：5 维度 code+name 成对输出；CASH 行仅 asset_class=现金、
  // 其余维度 NULL；IN_TRANSIT 在途行五维全 NULL
  asset_class_code?: string | null;
  asset_class_name?: string | null;
  region_code?: string | null;
  region_name?: string | null;
  style_code?: string | null;
  style_name?: string | null;
  size_code?: string | null;
  size_name?: string | null;
  segment_code?: string | null;
  segment_name?: string | null;
  // 读侧派生（issue #99）：daily_profit=当日收益（首个快照日 / IN_TRANSIT 在途行 → null）
  daily_profit?: number | null;
  /** 展示标签，不驱动取价（issue #228 起滞后估值看 nav_lag_days） */
  is_qdii?: boolean | null;
  /** issue #228：快照估值取价滞后交易日数（0=当日；N>0 时前端 tooltip 提示日收益滞后 N 天） */
  nav_lag_days?: number | null;
  snapshot_date: string;
  created_at?: string;
}

export interface PositionCreate {
  portfolio_code: string;
  product_code: string;
  market?: string;
  platform_code?: string;
  shares?: number;
  frozen_shares?: number;
  cost_price?: number;
  unit_price?: number;
  market_value?: number;
  cash_amount?: number;
  frozen_amount?: number;
  snapshot_date: string;
}

export interface PositionUpdate {
  platform_code?: string;
  shares?: number;
  frozen_shares?: number;
  cost_price?: number;
  unit_price?: number;
  market_value?: number;
  cash_amount?: number;
  frozen_amount?: number;
  snapshot_date?: string;
}
